README.md
<!---
This file is part of the sshilko/php-sql-mydb package.

(c) Sergei Shilko <contact@sshilko.com>

MIT License

For the full copyright and license information, please view the LICENSE
file that was distributed with this source code.

 @license https://opensource.org/licenses/mit-license.php MIT
-->
MyDb - Component
=================
<p align="center" cellpadding="0" cellspacing="0">
	<img src="https://img.shields.io/badge/status-active-success" alt="Project status - active">
	<a href="https://packagist.org/packages/sshilko/php-sql-mydb"><img src="https://poser.pugx.org/sshilko/php-sql-mydb/v/stable" alt="Latest Stable Version"></a>
	<a href="https://packagist.org/packages/sshilko/php-sql-mydb/stats"><img src="https://poser.pugx.org/sshilko/php-sql-mydb/downloads" alt="Total Downloads"></a>
	<a href="https://packagist.org/packages/sshilko/php-sql-mydb"><img src="https://poser.pugx.org/sshilko/php-sql-mydb/require/php" alt="PHP Required Version"></a>
	<a href="https://choosealicense.com/licenses/mit/"><img src="https://poser.pugx.org/sshilko/php-sql-mydb/license" alt="MIT License"></a>
    <a href="https://psalm.dev/docs/running_psalm/command_line_usage/#shepherd"><img src="https://shepherd.dev/github/sshilko/php-sql-mydb/coverage.svg" alt="Psalm Coverage"></a>
    <img src="https://hits.seeyoufarm.com/api/count/incr/badge.svg?url=https%3A%2F%2Fgithub.com%2Fsshilko%2Fphp-sql-mydb&count_bg=%2379C83D&title_bg=%23555555&icon=&icon_color=%23E7E7E7&title=hits&edge_flat=false"/>
    <br/>
    <img src="https://raw.githubusercontent.com/sshilko/php-sql-mydb/badges/phpunit-coverage-badge.svg" alt="PHPUnit coverage">
    <img src="https://raw.githubusercontent.com/sshilko/php-sql-mydb/badges/phpunit-coverage-badge-classes.svg" alt="PHPUnit classes coverage">
    <img src="https://raw.githubusercontent.com/sshilko/php-sql-mydb/badges/phpunit-coverage-badge-lines.svg" alt="PHPUnit lines coverage">
    <img src="https://raw.githubusercontent.com/sshilko/php-sql-mydb/badges/phpunit-coverage-badge-methods.svg" alt="PHPUnit methods coverage">
    <br/>
    <a href="https://github.com/sshilko/php-sql-mydb/actions/workflows/psalm.yml"><img src="https://github.com/sshilko/php-sql-mydb/actions/workflows/psalm.yml/badge.svg" alt="Psalm build"></a>
    <a href="https://github.com/sshilko/php-sql-mydb/actions/workflows/phan.yml"><img src="https://github.com/sshilko/php-sql-mydb/actions/workflows/phan.yml/badge.svg" alt="Phan build"></a>
    <a href="https://github.com/sshilko/php-sql-mydb/actions/workflows/phpunit.yml"><img src="https://github.com/sshilko/php-sql-mydb/actions/workflows/phpunit.yml/badge.svg" alt="PHPUnit build"></a>
    <a href="https://github.com/sshilko/php-sql-mydb/actions/workflows/phpmd.yml"><img src="https://github.com/sshilko/php-sql-mydb/actions/workflows/phpmd.yml/badge.svg" alt="PHPMd build"></a>
    <a href="https://github.com/sshilko/php-sql-mydb/actions/workflows/phpstan.yml"><img src="https://github.com/sshilko/php-sql-mydb/actions/workflows/phpstan.yml/badge.svg" alt="PHPStan build"></a>
    <a href="https://github.com/sshilko/php-sql-mydb/actions/workflows/phpcs.yml"><img src="https://github.com/sshilko/php-sql-mydb/actions/workflows/phpcs.yml/badge.svg" alt="PHPCodeSniffer build"></a>
</p>

Simple library to work with MySQL database.

### How this client helps you talk SQL to MySQL server

- Quality production defaults
  - TRADITIONAL [sql mode](https://dev.mysql.com/doc/refman/5.7/en/sql-mode.html#sqlmode_traditional)
  - autocommit 0, require commit to accept a transaction
  - automatically commit when gracefully shutting down
  - 90s client-side read-timeout for any query
  - 89s server-side SELECT query timeout
  - 5s client-side connect-timeout
  - respect php-fpm client disconnect and script execution shutdown
  - increased mysqlnd MYSQLI_OPT_NET_READ_BUFFER_SIZE and MYSQLI_OPT_NET_CMD_BUFFER_SIZE
  - automatic mysqlnd.net_read_timeout value
  - non-persistent connections
  - utf8mb4 charset
  - UTC timezone
  - 2hours non-interactive connection timeout
  - read-only transaction optimizations: isolation level READ COMMITTED, TRANSACTION READ ONLY
  - php error-reporting E_ALL & ~E_WARNING & ~E_NOTICE
  - Catch and report errors from mysqli function calls
    - MYSQLI_REPORT_ALL ^ MYSQLI_REPORT_STRICT ^ MYSQLI_REPORT_INDEX
  - Catch SIGTERM, SIGINT, SIGHUP termination signals
- Configurable connection retry
- Better memory usage - copy results from the mysqlnd buffer into the PHP variables MYSQLI_STORE_RESULT_COPY_DATA
- unit-tested, static analysed codebase source

#### What is the best use-case for this library

- MySQL server
- High performance, low-latency data-intensive applications
- API first applications with existing user-input marshalling
- No prepared statements requirement
- No 3rd party dependencies
- Minimum abstractions, no ORM no ActiveRecord patterns
- No run-time/compile-time steps
- Minimum code, maximum-performance

#### Why this library exists

* MySQL database is fast, reliable and scalable, php runtime is the same
* Value developers time and do not add complexity where possible
* Measure app's performance with real-world datasets and organic load
* Optimize for my use-case, focus on bottlenecks, remember that there is no ~~NoSQL~~ silver bullet
* Do not optimize early - resources like CPU, memory are cheap
* Focus on building architecture, learn from others and improve over time

##### Is it compatible?

- [+] PHP 7.4
- [+] MySQL <-> PHP via TCP
- [+] MySQL >=5.7.8
- [?] MySQL 8.0

##### Roadmap 2022-2023

- execute command/query via events, unit-tests against raw SQL generator and events objects
- 90% tests coverage
- usage example
- run phpunit tests against MySQL8
- run phpunit tests against PHP8
- Pluggable M.E.L.T (metrics, events, logs, traces)
- Packagist export & release

#### Installation

```
composer install --no-dev
```

#### Development setup

- Install [PHP](https://www.php.net/) & [Composer](https://getcomposer.org/) & [Docker Compose](https://docs.docker.com/compose/install/)
- Install [PHPStorm IDE](https://www.jetbrains.com/phpstorm/) with [PHP Inspections](https://github.com/kalessil/phpinspectionsea) or [VSCode IDE](https://code.visualstudio.com/)

```
docker-compose build
docker-compose up -d --no-build --wait
docker-compose exec -T app composer install --dev
...
docker-compose stop
```

#### Apply coding standards to modified files

`docker-compose exec app composer app-pre-commit`

```
docker-compose exec app composer app-phpcbf  &&
docker-compose exec app composer app-phpcs   &&
docker-compose exec app composer app-phpmd   &&
docker-compose exec app composer app-phpstan &&
docker-compose exec app composer app-psalm   && 
docker-compose exec app composer app-phan    &&
echo 'Looks good' 
```

#### Run [PHPUnit](https://phpunit.de) test suite

```
docker-compose exec app composer app-phpunit
```

#### Run PHP Code Beautifier & PHP [CodeSniffer](https://github.com/squizlabs/PHP_CodeSniffer) (all files)

```
docker-compose exec app composer app-phpcbf
docker-compose exec app composer app-phpcs
```

#### Run [Psalm](https://psalm.dev) - a static analysis tool for PHP (all files)

```
docker-compose exec app composer app-psalm
docker-compose exec app composer app-psalm-alter
docker-compose exec app composer app-psalm-taint
docker-compose exec app composer app-psalm-shepherd
```

#### Run [PHPStan](https://phpstan.org) - PHP Static Analysis Tool (all files)

`docker-compose exec app composer app-phpstan`

#### Run [PHPMD](https://phpmd.org) - PHP Mess Detector

```
docker-compose exec app composer app-phpmd
```

#### Run [Phan](https://github.com/phan/phan) - PHP Phan static analyzer

```
docker-compose exec app composer app-phan
```

#### Run [phpDocumentor](https://www.phpdoc.org) - [phpDocumentor](https://docs.phpdoc.org/3.0/guide/references/phpdoc/tags/)

```
docker-compose exec app composer app-phpdoc
```

#### Run [PHPCPD](https://github.com/sebastianbergmann/phpcpd) - PHPCPD Copy/Paste Detector (CPD) for PHP code

```
docker-compose exec app composer app-phpcpd
```

#### Run [Pdepend](https://pdepend.org) - PHP quality of design - extensibility, reusability and maintainability

```
docker-compose exec app composer app-pdepend
```


CHANGELOG.md
<!---
This file is part of the sshilko/php-sql-mydb package.

(c) Sergei Shilko <contact@sshilko.com>

MIT License

For the full copyright and license information, please view the LICENSE
file that was distributed with this source code.

 @license https://opensource.org/licenses/mit-license.php MIT
-->
CHANGELOG
=========

0.2
---

* Added PHP MessDetector, reduced psalm errorLevel from 5 to 3

0.1
---

* Initial release


src/sql/MydbMysqli.php
<?php
/**
 * This file is part of the sshilko/php-sql-mydb package.
 *
 * (c) Sergei Shilko <contact@sshilko.com>
 *
 * MIT License
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 * @license https://opensource.org/licenses/mit-license.php MIT
 */

declare(strict_types = 1);

namespace sql;

use mysqli;
use mysqli_result;
use sql\MydbMysqli\MydbMysqliResult;
use function array_merge;
use function array_values;
use function in_array;
use function max;
use function mysqli_init;
use function mysqli_query;
use function mysqli_report;
use function sprintf;
use const MYSQLI_ASYNC;
use const MYSQLI_INIT_COMMAND;
use const MYSQLI_OPT_CONNECT_TIMEOUT;
use const MYSQLI_OPT_NET_CMD_BUFFER_SIZE;
use const MYSQLI_OPT_NET_READ_BUFFER_SIZE;
use const MYSQLI_OPT_READ_TIMEOUT;
use const MYSQLI_REPORT_ALL;
use const MYSQLI_REPORT_INDEX;
use const MYSQLI_REPORT_STRICT;
use const MYSQLI_STORE_RESULT_COPY_DATA;
use const MYSQLI_TRANS_COR_NO_RELEASE;
use const MYSQLI_TRANS_COR_RELEASE;
use const MYSQLI_TRANS_START_READ_ONLY;
use const MYSQLI_TRANS_START_READ_WRITE;

/**
 * Facade for php mysqli extension
 *
 * @author Sergei Shilko <contact@sshilko.com>
 * @license https://opensource.org/licenses/mit-license.php MIT
 * @see https://github.com/sshilko/php-sql-mydb
 * @see https://www.php.net/manual/en/class.mysqli
 */
class MydbMysqli
{
    /**
     * Command to execute when connecting to MySQL server. Will automatically be re-executed when reconnecting.
     * @see https://www.php.net/manual/en/mysqli.constants.php
     */
    public const MYSQLI_INIT_COMMAND = MYSQLI_INIT_COMMAND;

    /**
     * Connect timeout in seconds
     * @see https://www.php.net/manual/en/mysqli.constants.php
     */
    public const MYSQLI_OPT_CONNECT_TIMEOUT = MYSQLI_OPT_CONNECT_TIMEOUT;

    /**
     * The size of the internal command/network buffer. Only valid for mysqlnd.
     * @see https://www.php.net/manual/en/mysqli.constants.php
     */
    public const MYSQLI_OPT_NET_CMD_BUFFER_SIZE = MYSQLI_OPT_NET_CMD_BUFFER_SIZE;

    /**
     * Maximum read chunk size in bytes when reading the body of a MySQL command packet. Only valid for mysqlnd.
     * @see https://www.php.net/manual/en/mysqli.constants.php
     */
    public const MYSQLI_OPT_NET_READ_BUFFER_SIZE = MYSQLI_OPT_NET_READ_BUFFER_SIZE;

    /**
     * Command execution result timeout in seconds. Available as of PHP 7.2.0.
     * @see https://www.php.net/manual/en/mysqli.constants.php
     */
    public const MYSQLI_OPT_READ_TIMEOUT = MYSQLI_OPT_READ_TIMEOUT;

    /**
     * Copy results from the internal mysqlnd buffer into the PHP variables fetched.
     * By default, mysqlnd will use a reference logic to avoid copying and duplicating results
     * held in memory. For certain result sets, for example, result sets with many small rows,
     * the copy approach can reduce the overall memory usage because PHP variables holding
     * results may be released earlier (available with mysqlnd only)
     * @see https://www.php.net/manual/en/mysqli.constants.php
     */
    public const MYSQLI_STORE_RESULT_COPY_DATA = MYSQLI_STORE_RESULT_COPY_DATA;

    /**
     * Appends "RELEASE" to mysqli_commit() or mysqli_rollback().
     * The RELEASE clause causes the server to disconnect the current client session
     * after terminating the current transaction
     *
     * @see https://dev.mysql.com/doc/refman/8.0/en/commit.html
     * @see https://www.php.net/manual/en/mysqli.constants.php
     */
    public const MYSQLI_TRANS_COR_RELEASE = MYSQLI_TRANS_COR_RELEASE;

    /**
     * Start the transaction as "START TRANSACTION READ ONLY" with mysqli_begin_transaction().
     *
     * @see https://www.php.net/manual/en/mysqli.constants.php
     */
    public const MYSQLI_TRANS_START_READ_ONLY = MYSQLI_TRANS_START_READ_ONLY;

    /**
     * Start the transaction as "START TRANSACTION READ WRITE" with mysqli_begin_transaction().
     * @see https://www.php.net/manual/en/mysqli.constants.php
     */
    public const MYSQLI_TRANS_START_READ_WRITE = MYSQLI_TRANS_START_READ_WRITE;

    /**
     * Appends "NO RELEASE" to mysqli_commit() or mysqli_rollback().
     * The NO RELEASE clause asks the server to not disconnect the current client session
     * after terminating the current transaction
     *
     * @see https://dev.mysql.com/doc/refman/8.0/en/commit.html
     * @see https://www.php.net/manual/en/mysqli.constants.php
     */
    public const MYSQLI_TRANS_COR_NO_RELEASE = MYSQLI_TRANS_COR_NO_RELEASE;

    /**
     * Set all options on (report all), report all warnings/errors.
     *
     * @see https://www.php.net/manual/en/mysqli.constants.php
     */
    public const MYSQLI_REPORT_ALL = MYSQLI_REPORT_ALL;

    /**
     * Report if no index or bad index was used in a query.
     *
     * @see https://www.php.net/manual/en/mysqli.constants.php
     */
    public const MYSQLI_REPORT_INDEX = MYSQLI_REPORT_INDEX;

    /**
     * Throw a mysqli_sql_exception for errors instead of warnings.
     *
     * @see https://www.php.net/manual/en/mysqli.constants.php
     */
    public const MYSQLI_REPORT_STRICT = MYSQLI_REPORT_STRICT;

    /**
     * Safe MySQL SQL_MODE
     * @see https://dev.mysql.com/doc/refman/5.7/en/sql-mode.html#sqlmode_traditional
     */
    protected const SQL_MODE = 'TRADITIONAL';

    /**
     * Mysqli instance
     * The mysqli extension allows you to access the functionality provided by MySQL 4.1 and above
     *
     * @see http://dev.mysql.com/doc/
     * @see https://www.php.net/manual/en/mysqli.overview.php
     * @see https://www.php.net/manual/en/intro.mysqli.php
     * @see https://www.php.net/manual/en/class.mysqli
     */
    protected ?mysqli $mysqli = null;

    /**
     * Flag indicating whether physical connection was established with remote server
     * Is connected to remove server
     */
    protected bool $isConnected = false;

    /**
     * Flag indicating whether SQL transaction was started
     * WARNING: best-effort, only guaranteed when library is used correctly
     * Is transaction started
     */
    protected bool $isTransaction = false;

    public function __construct(?mysqli $resource = null)
    {
        if (!$resource) {
            return;
        }

        $this->mysqli = $resource;
    }

    /**
     * Allocate mysqli resource instance, no physical connection to remote is done
     *
     */
    public function init(): bool
    {
        if (null !== $this->mysqli) {
            /**
             * Prevent zombie connections
             */
            return false;
        }

        $init = mysqli_init();
        if ($init) {
            $this->mysqli = $init;

            return true;
        }

        $this->mysqli = null;

        return false;
    }

    /**
     * Set various options that affect mysqli resource, before connection is established
     *
     * @see https://www.php.net/manual/en/mysqli.options.php
     * @throws MydbException\EnvironmentException
     */
    public function setTransportOptions(MydbOptions $options, MydbEnvironment $environment): bool
    {
        if (null === $this->mysqli) {
            return false;
        }

        $ignoreUserAbort = $environment->ignore_user_abort();
        $selectTimeout = $options->getServerSideSelectTimeout();

        /**
         * Prevent entry of invalid values such as those that are out of range, or NULL specified for NOT NULL columns
         * TRADITIONAL = strict mode
         *
         * @see https://dev.mysql.com/doc/refman/5.7/en/sql-mode.html#sqlmode_traditional
         */
        $mysqliInit = sprintf('SET SESSION sql_mode = %s', self::SQL_MODE);
        if ($ignoreUserAbort < 1) {
            $mysqliInit .= sprintf(', SESSION max_execution_time = %s', $selectTimeout * 10000);
        }

        $connectTimeout = $options->getConnectTimeout();
        $readTimeout = $options->getReadTimeout();
        $netReadTimeout = (string) (max($selectTimeout, $readTimeout) + $connectTimeout);

        return
            $environment->setMysqlndNetReadTimeout($netReadTimeout) &&
            $this->mysqli->options(self::MYSQLI_INIT_COMMAND, $mysqliInit) &&
            $this->mysqli->options(self::MYSQLI_OPT_CONNECT_TIMEOUT, $connectTimeout) &&
            $this->mysqli->options(self::MYSQLI_OPT_READ_TIMEOUT, $readTimeout) &&
            $this->mysqli->options(self::MYSQLI_OPT_NET_CMD_BUFFER_SIZE, $options->getNetworkBufferSize()) &&
            $this->mysqli->options(self::MYSQLI_OPT_NET_READ_BUFFER_SIZE, $options->getNetworkReadBuffer());
    }

    public function isTransactionOpen(): bool
    {
        /**
         * Ignore autocommit setting here
         */
        return $this->isTransaction;
    }

    public function isConnected(): bool
    {
        return $this->mysqli && $this->isConnected;
    }

    public function getMysqli(): ?mysqli
    {
        return $this->mysqli;
    }

    /**
     * @see https://www.php.net/manual/en/mysqli.real-query.php
     */
    public function realQuery(string $query): bool
    {
        if ($this->mysqli && $this->isConnected()) {
            return $this->mysqli->real_query($query);
        }

        return false;
    }

    /**
     * React to mysqli resource changes after query/command execution
     */
    public function readServerResponse(MydbEnvironment $environment): ?MydbMysqliResult
    {
        if ($this->mysqli && $this->isConnected()) {
            $events = [];
            $warnings = [];
            
            $result = $this->extractServerResponse($environment, $events);

            $fieldsCount = $this->getFieldCount();

            if ($this->getWarningCount() > 0) {
                $warnings = array_merge($warnings, $this->getWarnings());
            }
            if ($events) {
                $warnings = array_merge($warnings, array_values($events));
            }

            $response = new MydbMysqliResult($result, $warnings, $fieldsCount ?? 0);

            $error = $this->getError();
            if (null !== $error && '' !== $error) {
                $response->setErrorMessage($error);
            }

            $errno = $this->getErrNo();
            if ($errno > 0) {
                $response->setErrorNumber($errno);
            }

            return $response;
        }

        return null;
    }

    /**
     * @see https://www.php.net/manual/en/mysqli.real-escape-string.php
     */
    public function realEscapeString(string $string): ?string
    {
        if (!$this->mysqli || !$this->isConnected()) {
            return null;
        }

        return $this->mysqli->real_escape_string($string);
    }

    /**
     * @see https://www.php.net/manual/en/mysqli.begin-transaction.php
     * @psalm-suppress MissingParamType
     */
    public function beginTransactionReadwrite(): bool
    {
        if ($this->mysqli &&
            $this->isConnected() &&
            $this->mysqli->begin_transaction(self::MYSQLI_TRANS_START_READ_WRITE)) {
            $this->isTransaction = true;

            return true;
        }

        return false;
    }

    /**
     * @see https://www.php.net/manual/en/mysqli.begin-transaction.php
     * @psalm-suppress MissingParamType
     */
    public function beginTransactionReadonly(): bool
    {
        if ($this->mysqli &&
            $this->isConnected() &&
            $this->mysqli->begin_transaction(self::MYSQLI_TRANS_START_READ_ONLY)) {
            $this->isTransaction = true;

            return true;
        }

        return false;
    }

    /**
     * @see https://www.php.net/manual/en/mysqli.begin-transaction.php
     * @psalm-suppress MissingParamType
     */
    public function beginTransaction(): bool
    {
        if ($this->mysqli && $this->isConnected() && $this->mysqli->begin_transaction()) {
            $this->isTransaction = true;

            return true;
        }

        return false;
    }

    /**
     * @see https://www.php.net/manual/en/mysqli.rollback.php
     */
    public function rollback(): bool
    {
        if ($this->mysqli && $this->isConnected() && $this->mysqli->rollback(self::MYSQLI_TRANS_COR_NO_RELEASE)) {
            $this->isTransaction = false;

            return true;
        }

        return false;
    }

    /**
     * Commit transaction and release connection from server side
     */
    public function commitAndRelease(): bool
    {
        if ($this->mysqli && $this->isConnected() && $this->mysqli->commit(self::MYSQLI_TRANS_COR_RELEASE)) {
            $this->isTransaction = false;

            return true;
        }

        return false;
    }

    /**
     * @psalm-suppress MissingParamType
     */
    public function commit(): bool
    {
        if ($this->mysqli && $this->isConnected() && $this->mysqli->commit(self::MYSQLI_TRANS_COR_NO_RELEASE)) {
            $this->isTransaction = false;

            return true;
        }

        return false;
    }

    /**
     * @psalm-suppress MissingParamType
     */
    public function realConnect(...$args): bool
    {
        /* @phan-suppress-next-line PhanPartialTypeMismatchArgumentInternal */
        if ($this->mysqli && !$this->isConnected() && $this->mysqli->real_connect(...$args)) {
            $this->isConnected = true;

            return true;
        }

        return false;
    }

    public function mysqliReport(int $level): bool
    {
        return mysqli_report($level);
    }

    public function mysqliQueryAsync(string $command): void
    {
        $mysqli = $this->getMysqli();
        if (!$mysqli || !$this->isConnected()) {
            return;
        }

        mysqli_query($mysqli, $command, MYSQLI_ASYNC);
    }

    public function close(): bool
    {
        if ($this->mysqli) {
            if ($this->isConnected()) {
                /**
                 * Ignore close() success/failure
                 */
                $this->mysqli->close();
                $this->isConnected = false;
            }

            $this->mysqli = null;

            return true;
        }

        return false;
    }

    public function getConnectErrno(): ?int
    {
        return $this->mysqli
            ? $this->mysqli->connect_errno
            : null;
    }

    public function getConnectError(): ?string
    {
        return $this->mysqli
            ? $this->mysqli->connect_error
            : null;
    }

    public function getServerVersion(): ?int
    {
        return $this->mysqli
            ? $this->mysqli->server_version
            : null;
    }

    public function isServerGone(): bool
    {
        return in_array($this->getErrNo(), [2002, 2006], true);
    }

    public function getError(): ?string
    {
        return $this->mysqli
            ? $this->mysqli->error
            : null;
    }

    public function getErrNo(): ?int
    {
        return $this->mysqli
            ? $this->mysqli->errno
            : null;
    }

    public function getAffectedRows(): ?int
    {
        $rows = $this->mysqli
            ? $this->mysqli->affected_rows
            : null;
        if (0 === $rows || $rows > 0) {
            return $rows;
        }

        /**
         * mysqli_affected_rows
         * An integer greater than zero indicates the number of rows affected or retrieved.
         * Zero indicates that no records where updated for an UPDATE statement,
         * no rows matched the WHERE clause in the query or that no query has yet been executed.
         * -1 indicates that the query returned an error.
         */
        return null;
    }

    /**
     * @return int|string|null
     */
    public function getInsertId()
    {
        return $this->mysqli
            ? $this->mysqli->insert_id
            : null;
    }

    public function autocommit(bool $enable): bool
    {
        if ($this->mysqli && $this->mysqli->autocommit($enable)) {
            if ($enable) {
                /**
                 * Some statement implicitly commit transaction
                 * @see https://dev.mysql.com/doc/refman/8.0/en/implicit-commit.html
                 */
                $this->isTransaction = false;
            }

            return true;
        }

        return false;
    }

    /**
     * @phpcs:disable SlevomatCodingStandard.PHP.DisallowReference.DisallowedPassingByReference
     */
    protected function extractServerResponse(MydbEnvironment $environment, array &$events): ?mysqli_result
    {
        if (null === $this->mysqli) {
            return null;
        }

        $oldHandler = $environment->set_error_handler(static function (int $errno, string $error) use (&$events) {
            $events[$errno] = $error;

            return true;
        });

        $result = $this->mysqli->store_result(self::MYSQLI_STORE_RESULT_COPY_DATA);
        $environment->set_error_handler($oldHandler);

        if (false === $result) {
            return null;
        }

        return $result;
    }

    /**
     * Returns fields count caused by query execution
     * Requires store_result to be called first
     * @see mysqli::store_result()
     */
    protected function getFieldCount(): ?int
    {
        return $this->mysqli
            ? $this->mysqli->field_count
            : null;
    }

    /**
     * Returns warnings caused by query execution
     * Requires store_result to be called first
     * @see mysqli::store_result()
     */
    protected function getWarningCount(): ?int
    {
        return $this->mysqli
            ? $this->mysqli->warning_count
            : null;
    }

    protected function getWarnings(): array
    {
        if ($this->mysqli) {
            $warnings = $this->mysqli->get_warnings();
            $array = [];
            do {
                $array[] = $warnings->message;
            } while ($warnings->next());

            return $array;
        }

        return [];
    }
}
